### ISSUE TITLE
Segmentation Issues with Ovarian Pathology Images

### ISSUE THREAD

**Issue by anamika-yadav99** [microsoft/BiomedParse#25]
Dear Authors,
 
Thank you for releasing this model as open-source! It's a valuable resource for the community.
 
I am attempting to segment non-pathological (normal) ovarian pathology image (H&E stain) using the provided `example_prediction.py` script. My WSIs have a magnification of 20x and an MPP (Micrometers Per Pixel) of 0.4942 for both X and Y axes. I've created image patches of size 1024x1024 pixels for segmentation. My goal is to segment epithelial cells, connective tissue cells, neoplastic cells, stromal cells, and inflammatory cells.
 
However, the current segmentation results fail to segment the cell type of interest. I've tried adjusting the base magnification within the script from 20x to 40x, but this hasn't improved the segmentation.
 
Would you be able to offer any suggestions on the best parameters to use for segmenting ovarian pathology images with this model? I've attached a few example slides for your reference.
 
![Image](https://github.com/user-attachments/assets/076e166e-87d5-4e8e-896e-214c25323885)
![Image](https://github.com/user-attachments/assets/8aa35412-261d-4f19-89ac-190bab982554)
![Image](https://github.com/user-attachments/assets/17952a8b-5cf4-4a86-bd4e-43d64f7e3e17)

**Comment by davidsoong** [comment]
I also would like to say thank you to the authors for the great work and releasing the model.

Q: I found that the image magnification level for H&E images matters a lot, too. Will the model segment better if we add information like H&E, 20x in the prompt? check_mask_stats does not seem to like unsupported prompts/cell types when calculating the p-value.

Also, where can I find the ontology? In target_dist.json I only see 4 cell types supported for pathology.

  "Pathology": {
    "connective tissue cells": [
    "inflammatory cells": [
    "neoplastic cells": [
    "epithelial cells": 

Thanks.

**Comment by theodore-zhao** [comment]
> I also would like to say thank you to the authors for the great work and releasing the model.
> 
> Q: I found that the image magnification level for H&E images matters a lot, too. Will the model segment better if we add information like H&E, 20x in the prompt? check_mask_stats does not seem to like unsupported prompts/cell types when calculating the p-value.
> 
> Also, where can I find the ontology? In target_dist.json I only see 4 cell types supported for pathology.
> 
> "Pathology": { "connective tissue cells": [ "inflammatory cells": [ "neoplastic cells": [ "epithelial cells":
> 
> Thanks.

Hi there, thanks for your interest! To clarify, prompts like H&E or 20x won't help as they were not used in training the model. We are currently supporting p-value checking for a subset of the object types in the ontology, and will add more in the future. If you have any specific object types in mind, we can prioritize those. 

Thanks!

**Comment by WeisiX** [comment]
@theodore-zhao On the same note of pathology images, 
1) Would it help to fine-tune the model with user's own data of different resolutions (e.g., lower magnifications and more zoomed-out views) so that it could work better with larger images like the one shown in this thread? 
2) Noticed that the current training and inference setup is pretty strict with 1024x1024 image size. What would be the recommended way to make the model work with different image sizes, in both training (finetuning) and inference? Asking because the existing preprocessing of resizing will alter the image resolution, but usually in pathology maintaining the same resolution is key to recognizing certain tissue or cell structure.

Thank you for this amazing work!

**Comment by theodore-zhao** [comment]
> [@theodore-zhao](https://github.com/theodore-zhao) On the same note of pathology images,
> 
> 1. Would it help to fine-tune the model with user's own data of different resolutions (e.g., lower magnifications and more zoomed-out views) so that it could work better with larger images like the one shown in this thread?
> 2. Noticed that the current training and inference setup is pretty strict with 1024x1024 image size. What would be the recommended way to make the model work with different image sizes, in both training (finetuning) and inference? Asking because the existing preprocessing of resizing will alter the image resolution, but usually in pathology maintaining the same resolution is key to recognizing certain tissue or cell structure.
> 
> Thank you for this amazing work!

Yes, finetuning will definitely be helpful. The current model was trained with larger magnification. About image resolution, the FocalNet backbone should be already compatible with different resolutions, so you can try directly. There will be a distribution shift, but given that finetuning is expected this shouldn't be a big problem.

**Comment by WeisiX** [comment]
Thanks you for the timely response! @theodore-zhao 
What would be the minimum GPU hardware requirement for finetuning? I prepared a training set with 103 images of size 1024 by 1024 pixels, tried using one A100 GPU with 40GB memory, but running the train.sh script gave me a CUDA out of memory error. Would like to learn more about the GPU memory requirements and if multi-GPU is needed for fine-tuning.

**Comment by theodore-zhao** [comment]
> Thanks you for the timely response! [@theodore-zhao](https://github.com/theodore-zhao) What would be the minimum GPU hardware requirement for finetuning? I prepared a training set with 103 images of size 1024 by 1024 pixels, tried using one A100 GPU with 40GB memory, but running the train.sh script gave me a CUDA out of memory error. Would like to learn more about the GPU memory requirements and if multi-GPU is needed for fine-tuning.

We did use multiple GPUs for training, but one A100 40GB should also be able to fit the model and data. You can try reducing the batch size for training. Given that the data size is quite small, this shouldn't make the training too long.
